src/ProtoRoute/Message.hs

```diff
@@ -1,46 +1,159 @@
+-- wrong rules or mismatching field names
+data ValidationError = Incomplete String |
+                       WrongName  String |
+                       WrongType  String |
+                       Invalid    String deriving (Show, Eq, Ord)
 
-data Field = Field
-    { fieldName :: FieldName
-    , fieldValue :: FieldValue
-    } deriving (Show)
+incomplete :: ValidationError
+incomplete = Incomplete "Required field not filled"
 
+incorrectFieldNames :: ValidationError
+incorrectFieldNames = WrongName "Wrong field names"
+
+incorrectMsgNames :: ValidationError
+incorrectMsgNames = WrongName "Message names do not match"
+
+incorrectTypes :: ValidationError
+incorrectTypes = WrongType "There are mismatching types"
+
+invalid :: ValidationError
+invalid = Invalid "Your message is invalid"
+
+-- Shows value of a message field
 fromFV :: FieldValue -> String
 fromFV (FText txt) = fromTV txt
 fromFV (FInt num)  = fromTV num
 fromFV (FMsg msg)  = fromTV msg
 
+-- Shows value inside FieldValue constructor
 fromTV :: (Show a) => TValue a -> String
 fromTV (Req val)  = show val
 fromTV (Opt txt)  = show txt
 fromTV (Rep list) = show list
 fromTV Error = "Error"
 
+-- Checks if rules in msg match those in corresponding schema
+correct :: TValue a -> SchemaRule -> Bool
+correct (Req _) SReq = True
+correct (Opt _) SOpt = True
+correct (Rep _) SRep = True
+correct _ _ = False
+
+-- Checks if types are correct
+okTypes :: FieldValue -> SchemaRule -> SchemaType -> Bool
+okTypes (FText tvA) rA fTA = fTA == SText && correct tvA rA
+okTypes (FInt tvB) rB fTB  = fTB == SInt && correct tvB rB
+okTypes (FMsg tvC) rC fTC  = fTC == SMsg && correct tvC rC
+
+-- Verifies if there exists a validation error
+verify :: Bool -> ValidationError -> Either ValidationError ()
+verify True someError = Left someError
+verify False _ = Right ()
+
+-- Checks if a required field is filled
+reqValThere :: SchemaField -> MessageField -> Bool
+reqValThere (SchemaField SReq _ sfn) mf =
+    case mf of
+        (MessageField sfn _) -> True
+reqValThere _ _ = False
+
+-- Validates single field, checking msg attempt against schema
+validateField :: SchemaField -> MessageField -> Either ValidationError ()
+validateField (SchemaField sfr sft sfn) (MessageField mfn mfv)  = do
+    verify (mfn /= sfn) incorrectFieldNames
+    verify (not $ okTypes mfv sfr sft) incorrectTypes
+    verify (not $ reqValThere (SchemaField sfr sft sfn) (MessageField mfn mfv))
+           incomplete
+
+thereExistsSF :: SchemaField -> Bool
+thereExistsSF (SchemaField _ _ name) = unFieldName name /= ""
+
+thereExistsMF :: MessageField -> Bool
+thereExistsMF (MessageField name _) = unFieldName name /= ""
+
+validateMessage :: MessageSchema -> Message -> Either ValidationError ()
+validateMessage (MessageSchema sn sFields) (Message mn mFields) = do
+    verify (mn /= sn) incorrectMsgNames
+    let nameFieldMap = groupedFields
+    let results = DMS.elems nameFieldMap
```

Comment:
It's much cleaner if you move all variables under `where` and your `do`-blocks are just:

```haskell
verify foo error1
verify bar error2
```

Just how you did for `validateField`